### PR TITLE
Implement native Discord slash commands and some fixes

### PR DESCRIPTION
**Pull request by aaronson2012**
I implemented native discord slash commands and made various fixes (sorry didn't write them down so I forgot).

I can't think of any potential issues (with this PR, anyway).

### CODE REVIEW

File: dejavu_bot.py [hunk 1]
```diff
@@ -1,111 +1,152 @@
...
+    font = ImageFont.truetype("./fonts/Courier.ttf", size=14)  # debian path
 
     # Convert dict_items to a list
     color_items = list(ImageColor.colormap.items())
-    rand_color = choice(color_items)[0]
-    img = Image.new('RGB', (1000, 100), color=rand_color)
+
+    rand_color_name = choices(list(ImageColor.colormap.keys()))[0]
+    rand_color = ImageColor.getrgb(rand_color_name)
+    img = Image.new("RGB", (1000, 100), color=rand_color)
```
Comment (CollinEMac) [inline dejavu_bot.py]: I feel like this was more readable before. Can we change this back to using `choice`? I think this function should still work if we leave it alone.

File: dejavu_bot.py [hunk 2]
```diff
@@ -1,111 +1,152 @@
...
+        app_commands.Choice(
+            name="Retrieve a random message and put it in an image.", value="image"
+        ),
+        app_commands.Choice(
+            name="Retrieve a random message and you must guess who said it by mentioning them.",
+            value="whosaid",
+        ),
+    ]
+)
+async def dejavu(inter, choices: app_commands.Choice[str]):
```
Comment (CollinEMac) [inline dejavu_bot.py:70]: `choices` needs to have a different name because we import `random.choices`

File: dejavu_bot.py [hunk 3]
```diff
@@ -1,111 +1,152 @@
...
     """
-    On `/dejavu` grab a random message and post it
+    On `/dejavu text|image|whosaid` grab a random message and post it in the chosen format.
     """
-    channel = ctx.channel
+
+    # If a game of whosaid is already being played, do not continue.
+    if bot.whosaid["playing"] == True:
```
Comment (CollinEMac) [inline dejavu_bot.py]: Run pylint and you'll see 

```
dejavu_bot.py:76:7: C0121: Comparison 'bot.whosaid['playing'] == True' should be 'bot.whosaid['playing'] is True' if checking for the singleton value True, or 'bot.whosaid['playing']' if testing for truthiness (singleton-comparison)
```

https://pylint.readthedocs.io/en/latest/user_guide/messages/convention/singleton-comparison.html

File: dejavu_bot.py [hunk 4]
```diff
@@ -116,35 +157,69 @@ async def create_and_send_image(text, channel):
...
+    channel the game started in.
+
+    Then, the game logic runs.
     """
-    if message.author == bot.user:
+
+    if message.author.bot == True:
+        return
+    elif bot.whosaid["channel"] != message.channel.id:
         return
```
Comment (CollinEMac) [inline dejavu_bot.py]: These can be combined.

```suggestion
    if message.author.bot == True or bot.whosaid["channel"] != message.channel.id:
        return
```

File: dejavu_bot.py [hunk 5]
```diff
@@ -1,111 +1,152 @@
...
     delta = end - start
     int_delta = (delta.days * 24 * 60 * 60) + delta.seconds
     random_second = randrange(int_delta)
     return start + timedelta(seconds=random_second)
 
-async def create_and_send_response(rand_message, channel, arg):
-    """
-    Creates an image with the message author, content, and creation datetime
-    with a random colored background
-    """
```
Comment (CollinEMac) [inline dejavu_bot.py:71]: Let's keep this docstring.

File: dejavu_bot.py [hunk 6]
```diff
@@ -116,35 +157,69 @@ async def create_and_send_image(text, channel):
...
+        await message.reply(
+            "Wrong again! It was " + bot.whosaid["author"] + "! Game over!."
+        )
+        bot.whosaid["playing"] = False
+        bot.whosaid["second_chance"] = True
+
+
+# Sync slash command to Discord.
+@bot.event
+async def on_ready():
```
Comment (CollinEMac) [inline dejavu_bot.py:215]: Add a docstring, please.

File: dejavu_bot.py [hunk 7]
```diff
@@ -1,111 +1,152 @@
...
-        await channel.send(text)
+        await channel.send("Invalid Command.")
+
 
 async def create_and_send_image(text, channel):
     """
-    Handle the case where an image is requested
-    /dejavu image
+    Creates an image with the message author, content, and creation datetime
+    with a random colored background. TODO: Doesn't work right.
```
Comment (CollinEMac) [inline dejavu_bot.py]: I don't want to merge this until `create_and_send_image` works.

**Review by CollinEMac — CHANGES_REQUESTED** [review]
I think if you address these few things this should be g2g.

File: dejavu_bot.py (hunk 1 above)
Comment (aaronson2012) [reply dejavu_bot.py:149]: :+1:

File: dejavu_bot.py (hunk 2 above)
Comment (aaronson2012) [reply dejavu_bot.py:70]: :+1:

File: dejavu_bot.py (hunk 5 above)
Comment (aaronson2012) [reply dejavu_bot.py:71]: in the main dejavu function?

File: dejavu_bot.py (hunk 7 above)
Comment (aaronson2012) [reply dejavu_bot.py]: :+1:

File: dejavu_bot.py (hunk 6 above)
Comment (aaronson2012) [reply dejavu_bot.py:215]: :+1:

File: dejavu_bot.py (hunk 4 above)
Comment (aaronson2012) [reply dejavu_bot.py]: :+1:

File: dejavu_bot.py (hunk 3 above)
Comment (aaronson2012) [reply dejavu_bot.py]: :+1:

File: dejavu_bot.py (hunk 2 above)
Comment (aaronson2012) [reply dejavu_bot.py:70]: :+1:

File: dejavu_bot.py (hunk 1 above)
Comment (aaronson2012) [reply dejavu_bot.py]: :+1:

File: dejavu_bot.py [hunk 8]
```diff
@@ -1,109 +1,147 @@
...
-invoke with `/dejavu`
+Invoke with `/dejavu`
+Arguments: text, image, whosaid
 """
 
 import os
 from datetime import datetime, timedelta, timezone
 from io import BytesIO
-from random import choice, randrange
+from random import choices, randrange
```
Comment (CollinEMac) [inline dejavu_bot.py]:
```suggestion
from random import choice, randrange
```

File: dejavu_bot.py [hunk 9]
```diff
@@ -1,109 +1,147 @@
...
+        app_commands.Choice(
+            name="Retrieve a random message and put it in an image.", value="image"
+        ),
+        app_commands.Choice(
+            name="Retrieve a random message and you must guess who said it by mentioning them.",
+            value="whosaid",
+        ),
+    ]
+)
+async def dejavu(inter, arg: app_commands.Choice[str]):
```
Comment (CollinEMac) [inline dejavu_bot.py]:
```suggestion
async def dejavu(inter, choices: app_commands.Choice[str]):
```

File: dejavu_bot.py [hunk 10]
```diff
@@ -1,109 +1,147 @@
...
     end = datetime.utcnow().replace(tzinfo=timezone.utc)
     rand_datetime = get_rand_datetime(created_at, end)
 
+    # Fetch a message using the random datetime
     # limit=1 so we only get one message (we could change this later to add more?)
     async for rand_message in channel.history(limit=1, around=rand_datetime):
-        if rand_message.content != '':
-            await create_and_send_response(rand_message, channel, arg)
+        if rand_message.content != "":
+            await create_and_send_response(rand_message, channel, arg.value)
```
Comment (CollinEMac) [inline dejavu_bot.py]:
```suggestion
            await create_and_send_response(rand_message, channel, choices.value)
```

File: dejavu_bot.py [hunk 11]
```diff
@@ -116,35 +154,70 @@ async def create_and_send_image(text, channel):
...
+        bot.whosaid["second_chance"] = True
+
+
+# Sync slash command to Discord.
+@bot.event
+"""
+on_ready() syncs and updates the slash commands on the Discord server.
+"""
+async def on_ready():
+    await tree.sync()
```
Comment (CollinEMac) [inline dejavu_bot.py]:
```suggestion
@bot.event
async def on_ready():
    """
    on_ready() syncs and updates the slash commands on the Discord server.
    """
    await tree.sync()

```

**Review by CollinEMac — CHANGES_REQUESTED** [review]
I made these changes locally and it looks like it's working. Add these suggestions and :1st_place_medal:

**Review by aaronson2012 — COMMENTED** [review]
LGTM :+1:

**Review by CollinEMac — APPROVED** [review]

**Pull request merged**